### PR TITLE
Use UID for dockerfile to allow runasNonRoot to be used.

### PR DESCRIPTION
**Pull request by naemono**
See 
https://github.com/elastic/cloud-on-k8s/pull/6688#issuecomment-1514363649
and
https://github.com/elastic/cloud-on-k8s/issues/6740

Similar change was recently merged into [elasticsearch](https://github.com/elastic/elasticsearch/pull/95390) for the same issue.

This change will allow the beats container in Kubernetes to be ran with runAsNonRoot as it checks against the numeric UID, and not a name.

Without this change you encounter the following error:
```
  state:                                                                                                                                                                                                                                                                                                               │
│       waiting:                                                                                                                                                                                                                                                                                                           │
│         message: 'container has runAsNonRoot and image has non-numeric user (metricbeat),                                                                                                                                                                                                                                │
│           cannot verify user is non-root (pod: "elasticsearch-es-mixed-2_beats-3064(51b07a17-26c4-4301-9380-ea6fe3e75aca)",    
```

### CODE REVIEW

**Comment by belimawr** [comment]
LGTM, now it's just a matter of getting the tests passing.

The E2E failed to download the docker image, but it's available now. Let's see if re-running solves the problem.

**Comment by naemono** [comment]
> The E2E failed to download the docker image, but it's available now. Let's see if re-running solves the problem.

```
❯ docker pull docker.elastic.co/beats/filebeat:8.8.0-SNAPSHOT
Status: Downloaded newer image for docker.elastic.co/beats/filebeat:8.8.0-SNAPSHOT
docker.elastic.co/beats/filebeat:8.8.0-SNAPSHOT
```

Weird. Maybe I'll try one more time...

**Comment by naemono** [comment]
I manually told the job to re-run, and am watching the build...

**Comment by naemono** [comment]
@belimawr So I got the e2e tests to finish successfully twice now, but unfortunately as soon as it's done I've had to merge main into this branch, restarting the whole process which isn't a short process. It would be great to get this into the upcoming 8.8 stack release, as we're kinda on the fence in https://github.com/elastic/cloud-on-k8s/issues/6740 about whether we need to delay this until 8.9 and just remove `runAsNonRoot` setting in our stack monitoring.

Any thoughts on this being merged, and making it into the upcoming release?

**Comment by belimawr** [comment]
> @belimawr So I got the e2e tests to finish successfully twice now, but unfortunately as soon as it's done I've had to merge main into this branch, restarting the whole process which isn't a short process. It would be great to get this into the upcoming 8.8 stack release, as we're kinda on the fence in [elastic/cloud-on-k8s#6740](https://github.com/elastic/cloud-on-k8s/issues/6740) about whether we need to delay this until 8.9 and just remove `runAsNonRoot` setting in our stack monitoring.
> 
> Any thoughts on this being merged, and making it into the upcoming release?

Regarding merging `main` into this PR, it's not required. GitHub will ask you to do it, but it won't block the PR. So once the tests are passing, you can skip merging `main` onto this  PR.

Regarding getting it into the upcoming release, let me talk with the team. We've passed the FF already, so I can't make any promises.

**Comment by cmacknz** [comment]
There is another use of `USER root`, should that also be `USER 0`? It only applies when building the elastic-agent-complete docker image:

https://github.com/cmacknz/beats/blob/3806a9339c44d0e453f74760b2d21460405eac53/dev-tools/packaging/templates/docker/Dockerfile.tmpl#L156

**Comment by cmacknz** [comment]
Actually I now realize it's only the last `USER` that will matter since that's what the container will run as.

**Review by cmacknz — APPROVED** [review]

**Comment by cmacknz** [comment]
> Any thoughts on this being merged, and making it into the upcoming release?

The change seems benign and seems to be a bug fix where `runAsNonRoot` doesn't work only because of this missing change in Beats, I don't see a reason why we couldn't backport to 8.8.

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (cmacknz) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl:173]: Without us having a test to ensure `runAsNonRoot` keeps working there is a chance someone accidentally breaks this in the future.

Are there tests for this not breaking in ECK that we can rely on? Or should be adding a regression test in Beats to make sure this keeps working?

We have a k8s test pipeline stage that could conceivably include this check eventually:

https://github.com/cmacknz/beats/blob/3806a9339c44d0e453f74760b2d21460405eac53/Jenkinsfile#L321-L332

Right now it just makes sure the default k8s YAML deploys successfully.

Comment (naemono) [reply dev-tools/packaging/templates/docker/Dockerfile.tmpl:173]: Actually @cmacknz this is a catch-22. I can't get a test in place for this working until this feature is merged, and I know in what version of Beats this is functional.

Comment (cmacknz) [reply dev-tools/packaging/templates/docker/Dockerfile.tmpl:173]: Got it, thanks

**Pull request merged**
